### PR TITLE
Refactory all widgets.

### PR DESCRIPTION
**Pull request by terabytesoftw**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

- [x] Clean all code.
- [x] Use Html tags.
- [x] Fix issues infection.
- [x] Fix issues psalm.
- [x] Clean test, use `yiisoft/test-support`

### CODE REVIEW

File: src/Helper/HtmlForm.php [hunk 1]
```diff
@@ -128,9 +128,9 @@ public static function getAttributeName(string $attribute): string
      *
      * @throws InvalidArgumentException if the attribute name contains non-word characters.
      *
-     * @return array
+     * @return strings[] the attribute name, prefix and suffix.
```
Comment (samdark) [inline src/Helper/HtmlForm.php:131]:
```suggestion
     * @return string[] the attribute name, prefix and suffix.
```

File: src/Widget/Checkbox.php [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+/**
+ * Generates a checkbox tag together with a label for the given form attribute.
+ *
+ * This method will generate the "checked" tag attribute according to the form attribute value.
+ *
+ * @link https://www.w3.org/TR/2012/WD-html-markup-20120329/input.checkbox.html#input.checkbox
+ */
+final class Checkbox extends Widget
+{
+    private bool $unclosedByLabel = false;
```
Comment (samdark) [inline src/Widget/Checkbox.php:19]:
```suggestion
    private bool $enclosedByLabel = false;
```

File: src/Widget/Checkbox.php [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    private bool $uncheckValue = true;
```
Comment (samdark) [inline src/Widget/Checkbox.php:20]:
```suggestion
    private bool $forceUncheckedValue = true;
```

File: src/Widget/Checkbox.php [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        $new->attributes['required'] = $value;
+        return $new;
+    }
+
+    /**
+     * If the widget should be unclosed by label.
+     *
+     * @return static
+     */
+    public function unclosedByLabel(): self
```
Comment (samdark) [inline src/Widget/Checkbox.php:126]:
```suggestion
    public function enclosedByLabel(): self
```

File: src/Widget/Checkbox.php [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+     */
+    public function required(bool $value = true): self
+    {
+        $new = clone $this;
+        $new->attributes['required'] = $value;
+        return $new;
+    }
+
+    /**
+     * If the widget should be unclosed by label.
```
Comment (samdark) [inline src/Widget/Checkbox.php:122]:
```suggestion
     * If the widget should be enclosed by label.
```

File: src/Widget/Checkbox.php [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+    /**
+     * The value associated with the uncheck state of the checkbox.
+     *
+     * When this attribute is present, a hidden input will be generated so that if the checkbox is not checked and
+     * is submitted, the value of this attribute will still be submitted to the server via the hidden input.
+     *
+     * @return static
+     */
+    public function uncheckValue(): self
```
Comment (samdark) [inline src/Widget/Checkbox.php:141]:
```suggestion
    public function forceUncheckedValue(bool $value = true): self
```

File: src/Widget/Checkbox.php [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+     */
+    public function unclosedByLabel(): self
+    {
+        $new = clone $this;
+        $new->unclosedByLabel = true;
+        return $new;
+    }
+
+    /**
+     * The value associated with the uncheck state of the checkbox.
```
Comment (samdark) [inline src/Widget/Checkbox.php:134]:
```suggestion
     * Whether to generate hidden input for uncheck state of the checkbox.
```

File: src/Widget/CheckboxList.php [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+{
+    private array $containerAttributes = [];
+    private ?string $containerTag = 'div';
+    /** @psalm-var Closure(CheckboxItem):string|null */
+    private ?Closure $itemFormatter = null;
+    /** @var array<array-key, string> */
+    private array $items = [];
+    private array $itemsAttributes = [];
+    /** @var bool|float|int|string|Stringable|null */
+    private $uncheckValue = false;
```
Comment (samdark) [inline src/Widget/CheckboxList.php:27]:
```suggestion
    private $forceUncheckedValue = false;
```

File: src/Widget/CheckboxList.php [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+     * The value associated with the uncheck state of the checkboxlist.
+     *
+     * When this attribute is present, a hidden input will be generated so that if the checkboxlist is not checked and
+     * is submitted, the value of this attribute will still be submitted to the server via the hidden input.
+     *
+     * @param bool|float|int|string|Stringable|null $value
+     *
+     * @return static
+     */
+    public function uncheckValue($value): self
```
Comment (samdark) [inline src/Widget/CheckboxList.php:141]: Hm... it's not the same as in Checkbox?

Comment (terabytesoftw) [reply src/Widget/CheckboxList.php:141]: yes

File: src/Helper/HtmlForm.php (hunk 1 above)
Comment (terabytesoftw) [reply src/Helper/HtmlForm.php:131]: Done.

File: src/Widget/Checkbox.php (hunk 2 above)
Comment (terabytesoftw) [reply src/Widget/Checkbox.php:19]: done.

File: src/Widget/Checkbox.php (hunk 3 above)
Comment (terabytesoftw) [reply src/Widget/Checkbox.php:20]: done.

File: src/Widget/Checkbox.php (hunk 4 above)
Comment (terabytesoftw) [reply src/Widget/Checkbox.php:126]: done.

File: src/Widget/Checkbox.php (hunk 5 above)
Comment (terabytesoftw) [reply src/Widget/Checkbox.php:122]: done.

File: src/Widget/Checkbox.php (hunk 6 above)
Comment (terabytesoftw) [reply src/Widget/Checkbox.php:141]: done.

File: src/Widget/Checkbox.php (hunk 7 above)
Comment (terabytesoftw) [reply src/Widget/Checkbox.php:134]: done.

File: src/Widget/CheckboxList.php (hunk 8 above)
Comment (terabytesoftw) [reply src/Widget/CheckboxList.php:27]: done.

File: src/Widget/CheckboxList.php (hunk 9 above)
Comment (terabytesoftw) [reply src/Widget/CheckboxList.php:141]: done.

**Comment by terabytesoftw** [comment]
@samdark I can remove the hidden tag for all widgets, okay?

**Comment by samdark** [comment]
Yes.

**Comment by terabytesoftw** [comment]
@samdark `HasLengthHtmlOptions::class`, `MatchRegularExpressionHtmlOptions::class` to which widgets the validator attributes should be applied, add the `Email::class` and `Number::class` widgets, i have cleaned up all unnecessary classes and methods.

File: src/Widget/Form.php [hunk 10]
```diff
@@ -61,51 +73,195 @@ public function begin(): ?string
...
+     * Specifies whether the element represents an input control for which a UA is meant to store the value entered by
+     * the user (so that the UA can prefill the form later).
+     *
+     * @param string $value The value must be `on`,` off`.
+     *
+     * @return static
+     *
+     * @link https://www.w3.org/TR/html52/sec-forms.html#element-attrdef-autocompleteelements-autocomplete
+     */
+    public function autocomplete(string $value = 'on'): self
```
Comment (samdark) [inline src/Widget/Form.php:153]: How about bool?

File: src/Widget/Form.php [hunk 11]
```diff
@@ -61,51 +73,195 @@ public function begin(): ?string
...
-    public function options(array $value = []): self
+    /**
+     * The novalidate and formnovalidate content attributes are boolean attributes. If present, they indicate that the
+     * form is not to be validated during submission.
+     *
+     * @return static
+     *
+     * @link https://www.w3.org/TR/html52/sec-forms.html#element-attrdef-form-novalidate
+     */
+    public function noValidateHtml(): self
```
Comment (samdark) [inline src/Widget/Form.php:232]:
```suggestion
    public function withoutHtmlValidation(): self
```

Comment (samdark) [inline src/Widget/Form.php:232]:
```suggestion
    public function noHtmlValidatation(): self
```

File: src/Widget/Form.php [hunk 12]
```diff
@@ -61,51 +73,195 @@ public function begin(): ?string
...
+     * The target and formtarget content attributes, if specified, must have values that are valid browsing context
+     * names or keywords.
+     *
+     * @param string $value the target attribute value, for default its `_blank`.
+     *
+     * @return static
+     *
+     * @link https://www.w3.org/TR/html52/sec-forms.html#element-attrdef-form-target
+     */
+    public function target(string $value = '_blank'): self
```
Comment (samdark) [inline src/Widget/Form.php:249]: I'd not set a default here because `_blank` isn't really secure.

File: src/Widget/Label.php [hunk 13]
```diff
@@ -84,30 +24,60 @@ public function charset(string $value): self
...
-     * @param string $value
+     * @param string $value The label to be displayed.
      *
-     * @return self
+     * @return static
      *
      * Note that this will NOT be encoded.
-     * - If this is not set, {@see \Yiisoft\Form\FormModel::getAttributeLabel() will be called to get the label for
-     * display (after encoding).
+     * - If this is not set, {@see \Yii\Extension\Simple\Forms\BaseModel::getAttributeLabel() will be called to get the
```
Comment (samdark) [inline src/Widget/Label.php:46]: Huh?

File: src/Widget/Attribute/FieldAttribute.php [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+            if ($rule instanceof Number && $type === 'number') {
+                $attributes['max'] = $rule->getOptions()['max'];
+                $attributes['min'] = $rule->getOptions()['min'];
+            }
+            if ($rule instanceof Url && $type === 'url') {
+                $schemes = $rule->getOptions()['validSchemes'];
+                $pattern = $rule->getOptions()['pattern'];
+                $normalizePattern = str_replace('{schemes}', '(' . implode('|', $schemes) . ')', $pattern);
+                $attributes['pattern'] = Html::normalizeRegexpPattern($normalizePattern);
+            }
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/Attribute/FieldAttribute.php:183]: In my opinion, a widget should only accept the attributes set or allowed, previously you could add `HasLength::class` to a `RadioList::class` or `CheckBoxList::class`, now you cannot, you can combine the validators you want and that the `Input tag` allows according to https: // www. w3.org/, also you can in a custom rule you can always pass the `attribute` in `attributes` array.

File: src/Widget/Label.php (hunk 13 above)
Comment (terabytesoftw) [reply src/Widget/Label.php:46]: done.

File: src/Widget/Form.php (hunk 12 above)
Comment (terabytesoftw) [reply src/Widget/Form.php:249]: done.

File: src/Widget/Form.php (hunk 11 above)
Comment (terabytesoftw) [reply src/Widget/Form.php:232]: done.

Comment (terabytesoftw) [reply src/Widget/Form.php:232]: done.

File: src/Widget/Form.php (hunk 10 above)
Comment (terabytesoftw) [reply src/Widget/Form.php:153]: done.

File: src/Widget/Attribute/FieldAttribute.php [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+        $new->validClass = $value;
+        return $new;
+    }
+
+    private function addValidatorAttributeHtml(
+        FormModelInterface $formModel,
+        string $attribute,
+        array $attributes
+    ): array {
+        $haslengthTypes = ['email', 'password', 'tel', 'text', 'textarea', 'url'];
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/Attribute/FieldAttribute.php:157]: done.

File: src/FormModel.php [hunk 16]
```diff
@@ -30,71 +30,63 @@
...
     }
 
     public function getAttributeLabel(string $attribute): string
     {
-        if (array_key_exists($attribute, $this->attributesLabels)) {
-            return $this->attributesLabels[$attribute];
+        $label = $this->generateAttributeLabel($attribute);
+
+        /** @var array */
+        $labels = method_exists($this, 'getAttributeLabels') ? $this->getAttributeLabels() : [];
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/FormModel.php]: done.

File: src/FormModel.php [hunk 17]
```diff
@@ -30,71 +30,63 @@
...
 
-    public function getAttributeLabels(): array
-    {
-        return [];
-    }
-
     public function getAttributeHint(string $attribute): string
     {
+        /** @var array */
+        $attributeHints = method_exists($this, 'getAttributeHints') ? $this->getAttributeHints() : [];
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/FormModel.php]: done.

**Comment by terabytesoftw** [comment]
Please don't merge, i need to add `Button::class`, `ResetButton::class`, `SubmitButton::class`, `Color::class`, `Image::class`, `Time::class`, `Week::class`, `Range::class`.

File: composer.json [hunk 18]
```diff
@@ -19,7 +19,7 @@
     "prefer-stable": true,
     "require": {
         "php": "^7.4|^8.0",
-        "yiisoft/html": "^1.0",
+        "yiisoft/html": "dev-master as 1.2.0",
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply composer.json:22]: done.

File: src/Widget/Email.php [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+     *
+     * Default value is 4.
+     *
+     * @param int $value
+     *
+     * @return static
+     *
+     * @link https://www.w3.org/TR/2012/WD-html-markup-20120329/input.email.html#input.email.attrs.size
+     */
+    public function size(int $value = 4): self
```
Comment (vjik) [inline src/Widget/Email.php:124]: Why this default value?

Comment (terabytesoftw) [reply src/Widget/Email.php:124]: done.

File: src/Widget/Email.php [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+     * This must be an non-negative integer value smaller than or equal to the value specified by maxlength.
+     * If no minlength is specified, or an invalid value is specified, the text input has no minimum length.
+     *
+     * @param int $value
+     *
+     * @return static
+     *
+     * @link https://html.spec.whatwg.org/multipage/input.html#attr-input-minlength
+     */
+    public function minlength(int $value): self
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/Email.php:56]: done.

File: src/Widget/Email.php [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+     *
+     * If no maxlength is specified, or an invalid value is specified, the tag input has no maximum length.
+     *
+     * @param int $value Positive integer.
+     *
+     * @return static
+     *
+     * @link https://www.w3.org/TR/2012/WD-html-markup-20120329/input.email.html#input.email.attrs.maxlength
+     */
+    public function maxlength(int $value): self
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/Email.php:37]: done.

File: src/Widget/Attribute/ModelAttribute.php [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+
+declare(strict_types=1);
+
+namespace Yiisoft\Form\Widget\Attribute;
+
+use Yiisoft\Form\FormModelInterface;
+use Yiisoft\Form\Helper\HtmlForm;
+use Yiisoft\Html\Html;
+
+trait ModelAttribute
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/Attribute/ModelAttribute.php:11]: done.

File: src/FormModel.php [hunk 23]
```diff
@@ -134,6 +158,7 @@ public function getErrorSummary(bool $showAllErrors): array
         $lines = [];
         $errors = $showAllErrors ? $this->getErrors() : [$this->getFirstErrors()];
 
+        /** @var array $error */
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/FormModel.php:161]: done.

File: src/Widget/Attribute/DateAttribute.php [hunk 24]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Form\Widget\Attribute;
+
+trait DateAttribute
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/Attribute/DateAttribute.php:7]: done.

File: src/Widget/CheckboxList.php [hunk 25]
```diff
@@ -0,0 +1,235 @@
...
+    }
+
+    /**
+     * The HTML code that separates items.
+     *
+     * @param string $value
+     *
+     * @return static
+     */
+    public function separator(string $value): self
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/CheckboxList.php:186]: done.

File: src/Widget/CheckboxList.php [hunk 26]
```diff
@@ -0,0 +1,235 @@
...
+     * function ($index, $label, $name, $checked, $value)
+     * ```
+     *
+     * @param Closure $value
+     *
+     * @return static
+     *
+     * @psalm-param Closure(CheckboxItem):string|null $value
+     */
+    public function itemsFormatter(?Closure $value): self
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/CheckboxList.php:157]: done.

File: src/Widget/CheckboxList.php [hunk 27]
```diff
@@ -0,0 +1,235 @@
...
+    /**
+     * The items attributes for generating the list of checkboxes tag using {@see CheckBoxList}.
+     *
+     * @param array $value
+     *
+     * @return static
+     *
+     * {@see \Yiisoft\Html\Html::renderTagAttributes()} for details on how attributes are being rendered.
+     */
+    public function itemsAttributes(array $value = []): self
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/CheckboxList.php:134]: done.

File: src/Widget/CheckboxList.php [hunk 28]
```diff
@@ -0,0 +1,235 @@
...
+     *
+     * The array keys are the list of checkboxes values, and the array values are the corresponding labels.
+     *
+     * Note that the labels will NOT be HTML-encoded, while the values will.
+     *
+     * @param array<array-key, string> $value
+     *
+     * @return static
+     */
+    public function items(array $value = []): self
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/CheckboxList.php:118]: done.

File: src/Widget/CheckboxList.php [hunk 29]
```diff
@@ -0,0 +1,235 @@
...
+    /**
+     * The specified attributes for items.
+     *
+     * @param array $value
+     *
+     * @return static
+     *
+     * @psalm-param array<array-key, array<array-key, mixed>> $value
+     */
+    public function individualItemsAttributes(array $value = []): self
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/CheckboxList.php:100]: done.

File: src/Widget/CheckboxList.php [hunk 30]
```diff
@@ -0,0 +1,235 @@
...
+    /**
+     * The container attributes for generating the list of checkboxes tag using {@see CheckBoxList}.
+     *
+     * @param array $value
+     *
+     * @return static
+     *
+     * {@see \Yiisoft\Html\Html::renderTagAttributes()} for details on how attributes are being rendered.
+     */
+    public function containerAttributes(array $value): self
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/CheckboxList.php:48]: done.

File: src/Widget/Checkbox.php [hunk 31]
```diff
@@ -0,0 +1,116 @@
...
+     *
+     * Do not set this option unless you set the "label" attributes.
+     *
+     * @param array $value
+     *
+     * @return static
+     *
+     * {@see \Yiisoft\Html\Html::renderTagAttributes()} for details on how attributes are being rendered.
+     */
+    public function labelAttributes(array $value = []): self
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/Checkbox.php:77]: done.

File: src/Widget/Attribute/WithoutModelAttribute.php [hunk 32]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    protected function getId(): string
+    {
+        return $this->id = $this->id !== '' ? $this->id : $this->autoIdPrefix . self::$counter++;
+    }
+
+    protected function getName(): string
+    {
+        return $this->name = $this->name !== '' ? $this->name : $this->id;
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/Attribute/WithoutModelAttribute.php:113]: done.

File: src/Widget/Attribute/WithoutModelAttribute.php [hunk 33]
```diff
@@ -0,0 +1,115 @@
...
+     * @param int $value
+     */
+    public static function counter(int $value): void
+    {
+        self::$counter = $value;
+    }
+
+    protected function getId(): string
+    {
+        return $this->id = $this->id !== '' ? $this->id : $this->autoIdPrefix . self::$counter++;
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/Attribute/WithoutModelAttribute.php:108]: done.

File: src/Widget/Attribute/DateAttribute.php [hunk 34]
```diff
@@ -0,0 +1,40 @@
...
+    /**
+     * The latest acceptable date.
+     *
+     * @param string $value
+     *
+     * @return static
+     *
+     * @link https://html.spec.whatwg.org/multipage/input.html#attr-input-max
+     */
+    public function max(string $value): self
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/Attribute/DateAttribute.php:34]: Why null ?

File: src/Widget/Attribute/DateAttribute.php [hunk 35]
```diff
@@ -0,0 +1,40 @@
...
+    /**
+     * The earliest acceptable date.
+     *
+     * @param string $value
+     *
+     * @return static
+     *
+     * @link https://html.spec.whatwg.org/multipage/input.html#attr-input-min
+     */
+    public function min(string $value): self
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/Attribute/DateAttribute.php:18]: Why null ?

File: src/Helper/HtmlForm.php [hunk 36]
```diff
@@ -5,13 +5,59 @@
...
+     *
+     * @throws InvalidArgumentException if the attribute name contains non-word characters.
+     *
+     * @return string the attribute name without prefix and suffix.
+     *
+     * @see static::parseAttribute()
+     */
+    public static function getAttributeName(string $attribute): string
+    {
+        return self::parseAttribute($attribute)['name'];
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Helper/HtmlForm.php:58]: Done.

File: src/Helper/HtmlForm.php [hunk 37]
```diff
@@ -5,13 +5,59 @@
...
 
 /**
  * Form-related HTML tag generation
  */
 final class HtmlForm
 {
+    /**
+     * Return the attribute hint for the model.
+     *
+     * @param FormModelInterface|null $form the form object.
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Helper/HtmlForm.php:19]: It shouldn't be null, however the psalm gives an error if I don't initialize it.

File: src/FormModelInterface.php [hunk 38]
```diff
@@ -39,51 +31,44 @@ public function getAttributeLabels(): array;
...
      * @param string $attribute the attribute name.
      *
-     * @return string the attribute hint.
+     * @return string the attribute placeholder.
      *
-     * {@see getAttributeHints()}
+     * {@see getAttributePlaceHolders()}
      */
-    public function getAttributeHint(string $attribute): string;
+    public function getAttributePlaceHolder(string $attribute): string;
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/FormModelInterface.php:53]: Done.

File: src/FormModelInterface.php [hunk 39]
```diff
@@ -39,51 +31,44 @@ public function getAttributeLabels(): array;
...
      *
-     * Note, in order to inherit hints defined in the parent class, a child class needs to merge the parent hints with
-     * child hints using functions such as `array_merge()`.
+     * @return string the attribute hint.
      *
-     * @return array attribute hints (name => hint)
+     * {@see getAttributeHints()}
      */
-    public function getAttributeHints(): array;
+    public function getAttributeHint(string $attribute): string;
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/FormModelInterface.php:42]: done.

File: src/FormModel.php [hunk 40]
```diff
@@ -245,11 +278,9 @@ public function getRules(): array
         return [];
     }
 
-    /**
-     * @param string[][] $items
-     */
     private function addErrors(array $items): void
     {
+        /** @var array<string, array<array-key, string>> $items */
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/FormModel.php:283]: done.

File: src/FormModel.php [hunk 41]
```diff
@@ -190,16 +215,23 @@ public function load(array $data, ?string $formName = null): bool
         if ($scope === '' && !empty($data)) {
             $values = $data;
         } elseif (isset($data[$scope])) {
+            /** @var mixed */
             $values = $data[$scope];
         }
 
+        /** @var array<string, scalar|Stringable|null> $values */
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/FormModel.php:222]: done.

File: src/FormModel.php [hunk 42]
```diff
@@ -30,71 +30,95 @@
 abstract class FormModel implements FormModelInterface, PostValidationHookInterface, RulesProviderInterface
 {
     private array $attributes;
-    private array $attributesLabels;
+    /** @var array<string, array<array-key, string>> */
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/FormModel.php:33]: done.

File: src/FormModel.php [hunk 43]
```diff
@@ -30,71 +30,95 @@
...
     {
         return [];
     }
 
-    public function getAttributeHint(string $attribute): string
+    public function getAttributePlaceHolder(string $attribute): string
     {
+        $attributePlaceHolders = $this->getAttributePlaceHolders();
+
+        /** @var string */
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/FormModel.php:97]: done.

File: src/FormModel.php [hunk 44]
```diff
@@ -30,71 +30,95 @@
...
+    }
+
+    public function getAttributeLabel(string $attribute): string
+    {
+        $label = $this->generateAttributeLabel($attribute);
+
+        $labels = $this->getAttributeLabels();
+
+        if (array_key_exists($attribute, $labels)) {
+            /** @var string */
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/FormModel.php:73]: done.

File: src/FormModel.php [hunk 45]
```diff
@@ -30,71 +30,95 @@
...
+        $attributeHints = $this->getAttributeHints();
 
-        foreach ($validators as $validator) {
-            if ($validator instanceof Required) {
-                return true;
-            }
-            if ($validator instanceof HtmlOptionsProvider && (bool)($validator->getHtmlOptions()['required'] ?? false)) {
-                return true;
-            }
+        /** @var string */
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/FormModel.php:47]: done.

File: src/FormModel.php [hunk 46]
```diff
@@ -330,6 +361,13 @@ private function generateAttributeLabel(string $name): string
         );
     }
 
+    /**
+     * @return iterable|scalar|Stringable|null
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/FormModel.php:365]: done.

File: src/Helper/HtmlForm.php (hunk 37 above)
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Helper/HtmlForm.php]: fixed.

File: src/FormModel.php [hunk 47]
```diff
@@ -30,71 +30,95 @@
...
+        /** @var string */
+        $placeHolder = $attributePlaceHolders[$attribute] ?? '';
+
         [$attribute, $nested] = $this->getNestedAttribute($attribute);
+
         if ($nested !== null) {
-            return $this->readProperty($attribute)->getAttributeHint($nested);
+            /** @var FormModelInterface $attributeNestedValue */
+            $attributeNestedValue = $this->getAttributeValue($attribute);
+            $placeHolder = $attributeNestedValue->getAttributePlaceHolder($nested);
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/FormModel.php:103]: https://github.com/terabytesoftw/yii-form/blob/93d492c60a59023becf2136b46252f6e92c74817/src/FormModel.php#L449

File: src/FormMetadataInterface.php [hunk 48]
```diff
@@ -0,0 +1,79 @@
...
+    public function getAttributeHints(): array;
+
+    /**
+     * Returns the text placeholder for the specified attribute.
+     *
+     * @param string $attribute the attribute name.
+     *
+     * @return string the attribute placeholder.
+     */
+    public function getAttributePlaceHolder(string $attribute): string;
```
Comment (vjik) [inline src/FormMetadataInterface.php:71]:
```suggestion
    public function getAttributePlaceholder(string $attribute): string;
```

File: src/FormMetadataInterface.php [hunk 49]
```diff
@@ -0,0 +1,79 @@
...
+
+    /**
+     * Returns the attribute placeholders.
+     *
+     * @return array attribute placeholder (name => placeholder)
+     */
+    public function getAttributePlaceHolders(): array;
```
Comment (vjik) [inline src/FormMetadataInterface.php:78]:
```suggestion
    public function getAttributePlaceholders(): array;
```

Comment (terabytesoftw) [reply src/FormMetadataInterface.php]: done.

File: src/FormMetadataInterface.php (hunk 48 above)
Comment (terabytesoftw) [reply src/FormMetadataInterface.php]: done.

File: src/Widget/DateTime.php [hunk 50]
```diff
@@ -0,0 +1,49 @@
...
+
+use InvalidArgumentException;
+use Yiisoft\Form\Helper\HtmlForm;
+use Yiisoft\Form\Widget\Attribute\CommonAttribute;
+use Yiisoft\Form\Widget\Attribute\DateAttributes;
+use Yiisoft\Form\Widget\Attribute\ModelAttributes;
+use Yiisoft\Html\Tag\Input;
+use Yiisoft\Widget\Widget;
+
+/*
```
Comment (vjik) [inline src/Widget/DateTime.php:15]:
```suggestion
/**
```

File: src/Widget/DateTimeLocal.php [hunk 51]
```diff
@@ -0,0 +1,51 @@
...
+
+use InvalidArgumentException;
+use Yiisoft\Form\Helper\HtmlForm;
+use Yiisoft\Form\Widget\Attribute\CommonAttribute;
+use Yiisoft\Form\Widget\Attribute\DateAttributes;
+use Yiisoft\Form\Widget\Attribute\ModelAttributes;
+use Yiisoft\Html\Tag\Input;
+use Yiisoft\Widget\Widget;
+
+/*
```
Comment (vjik) [inline src/Widget/DateTimeLocal.php:15]:
```suggestion
/**
```

File: src/Widget/DateTime.php (hunk 50 above)
Comment (terabytesoftw) [reply src/Widget/DateTime.php]: done.

File: src/Widget/Attribute/DateAttribute.php (hunk 35 above)
[the remark this answers is not in the document]
Comment (vjik) [reply src/Widget/Attribute/DateAttribute.php]: I can use widget in views with variable `?string $minDate` . Will be convenient set this: `$widget->min($minDate)`.

File: src/Widget/DateTimeLocal.php (hunk 51 above)
Comment (terabytesoftw) [reply src/Widget/DateTimeLocal.php:15]: done.

**Comment by terabytesoftw** [comment]
Done https://github.com/yiisoft/form/pull/86#discussion_r698746443

File: src/Widget/Attribute/CommonAttribute.php [hunk 52]
```diff
@@ -2,61 +2,24 @@
...
 declare(strict_types=1);
 
-namespace Yiisoft\Form\Widget;
+namespace Yiisoft\Form\Widget\Attribute;
 
-use Yiisoft\Form\FormModelInterface;
-use Yiisoft\Widget\Widget;
-
-final class DatePicker extends Widget
+trait CommonAttribute
```
[the remark this answers is not in the document]
Comment (terabytesoftw) [reply src/Widget/Attribute/CommonAttribute.php]: done.

File: src/Widget/Attribute/CommonAttributes.php [hunk 53]
```diff
@@ -71,54 +34,62 @@ public function autofocus(bool $value = true): self
...
+     * @return static
+     *
+     * @link https://www.w3.org/TR/html52/sec-forms.html#the-required-attribute
      */
-    public function max(string $value): self
+    public function required(bool $value = true): self
     {
         $new = clone $this;
-        $new->options['max'] = $value;
+        $new->attributes['required'] = true;
```
Comment (samdark) [inline src/Widget/Attribute/CommonAttributes.php:77]:
```suggestion
        $new->attributes['required'] = $value;
```

or remove an argument.

**Review by samdark — APPROVED** [review]

File: src/Widget/Attribute/CommonAttributes.php (hunk 53 above)
Comment (terabytesoftw) [reply src/Widget/Attribute/CommonAttributes.php]: done.

**Comment by samdark** [comment]
@terabytesoftw how about merging it now and continuing with docs etc. in separate PRs?

**Comment by terabytesoftw** [comment]
> @terabytesoftw how about merging it now and continuing with docs etc. in separate PRs?

Yes 👍

**Comment by samdark** [comment]
👍

**Comment by samdark** [comment]
@terabytesoftw done. Please close all the issues fixed with it. Thanks.

**Pull request merged**